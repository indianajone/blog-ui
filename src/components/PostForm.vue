<template>
    <div class="section">
        <h1 class="title is-size-3">{{ pageName }} Post</h1>
        <form class="form">
            <div class="field">
                <div class="control">
                <input class="input" type="text" name="title" placeholder="What's it about..." v-model="title" required>
                </div>
            </div>
             <div class="field">
                <div class="control">
                    <div class="file has-name is-fullwidth">
                        <label class="file-label">
                            <input class="file-input" type="file" name="image" @change="onFileChange">
                            <span class="file-cta">
                                <span class="file-icon">
                                    <i class="fa fa-upload"></i>
                                </span>
                                <span class="file-label">
                                     Choose a thumbnial...
                                </span>
                            </span>
                            <span class="file-name">
                                {{ image ? image.name : '' }}
                            </span>
                        </label>
                    </div>
                </div>
             </div>
            <div class="field">
                <div class="control">
                <vue-html5-editor :content="body" :height="500" @change="onContentUpdate"></vue-html5-editor>
                </div>
            </div>
            <div class="field is-grouped">
                <div class="control">
                <button class="button is-link" type="button" @click="submit">Post</button>
                </div>
                <div class="control">
                <button class="button is-text" type="button" @click="clear">Cancel</button>
                </div>
            </div>
        </form>
    </div>
</template>

<script>
    export default {
        name: 'PostForm',
        data() {
            return {
                title: null,
                image: null,
                body: 'Start your story...'
            };
        },
        computed: {
            isEditMode() {
                return this.$route.name === 'EditPost';
            },
            pageName () {
                return this.isEditMode ? 'Edit' : 'Create';
            }
        },
        mounted() {
            if (this.isEditMode) {
                this.fetch();
            }
        },
        methods: {
            async fetch() {
                try {
                    const { id } = this.$route.params;
                    const { data } = await this.$http.get(`/posts/${id}`);
                    this.title = data.data.title;
                    this.body = data.data.body;
                } catch (e) {
                    console.error(e);
                }
            },
            onContentUpdate(content) {
                this.body = content;
            },
            onFileChange(e) {
                const files = e.target.files || e.dataTransfer.files;
                if (files.length) {
                    this.image = files[0];
                }
            },
            async createPost() {
                const formData = new FormData();
                formData.append('title', this.title);
                formData.append('body', this.body);

                if (this.image) {
                    formData.append('image', this.image);
                }

                try {
                    await this.$http.post('/posts', formData, {
                        headers: {
                            'Content-Type': 'multipart/form-data'
                        }
                    });
                } catch (e) {
                    console.error(e);
                }
            },
            submit() {
                this.createPost().then(() => {
                    this.$router.go('/');
                });
            },
            clear() {
                this.title = null;
                this.body = 'Start your story...';
                this.image = null;
            }
        }
    };
</script>

<style lang="scss">
    $fa-font-path: "~font-awesome/fonts";
    @import '~font-awesome/scss/font-awesome';
</style>


