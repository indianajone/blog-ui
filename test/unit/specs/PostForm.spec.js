import { mount } from 'vue-test-utils';
import PostForm from '@/components/PostForm';
import { createRenderer } from 'vue-server-renderer';

describe('PostForm.vue', () => {

    it('should render as create mode.', () => {
        const wrapper = mount(PostForm, {
            mocks: {
                $route: {
                    name: 'CreatePost'
                }
            }
        });

        expect(wrapper.find('.title').text()).toBe('Create Post');

        const renderer = createRenderer();
        renderer.renderToString(wrapper.vm, (err, str) => {

            if (err) {
                throw new Error(err);
            }
            expect(str).toMatchSnapshot();

        });
    });

    it('should render as edit mode.', () => {
        const wrapper = mount(PostForm, {
            mocks: {
                $route: {
                    name: 'EditPost',
                    params: { id: '59eccf0e4e8b8f72705ac87d' }
                },
                $http: {
                    defaults: { baseURL: 'localhost' },
                    get() {
                        return { data: {
                            data: {
                                title: 'New Post',
                                body: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
                                _id: '59eccf0e4e8b8f72705ac87d',
                                createdAt: '2017-10-02T17:02:06.051Z',
                                type: 'text'
                            }
                        } };
                    }
                }
            }
        });

        expect(wrapper.find('.title').text()).toBe('Edit Post');

        const renderer = createRenderer();
        renderer.renderToString(wrapper.vm, (err, str) => {

            if (err) {
                throw new Error(err);
            }
            expect(str).toMatchSnapshot();

        });
    });

    it('should submit a post to the server.', () => {
        // eslint-disable-next-line no-unused-vars
        const postMock = jest.fn((...args) => Promise.resolve());
        const wrapper = mount(PostForm, {
            mocks: {
                $route: {
                    name: 'CreatePost'
                },
                $router: {
                    push: () => Promise.resolve()
                },
                $http: {
                    post: postMock
                }

            }
        });

        wrapper.setData('title', 'HelloWorld');
        wrapper.setData('body', 'Write someting great.');

        wrapper.find('#submit-btn').trigger('click');

        const [args] = postMock.mock.calls;
        expect(args[0]).toBe('/posts');
        expect(args[1]).toBeInstanceOf(FormData);
        expect(args[2].headers['Content-Type']).toBe('multipart/form-data');
    });

});
