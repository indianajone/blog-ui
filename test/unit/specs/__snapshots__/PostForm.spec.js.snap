// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm.vue should render as create mode. 1`] = `
<div class="section">
    <h1 class="title is-size-3">Create Post</h1>
    <form class="form">
        <div class="field">
            <div class="control"><input type="text" name="title" placeholder="What's it about..." required="required" class="input"></div>
        </div>
        <div class="field">
            <div class="control">
                <div class="file has-name is-fullwidth"><label class="file-label"><input type="file" name="image" class="file-input"> <span class="file-cta"><span class="file-icon"><i class="fa fa-upload"></i></span> <span class="file-label">
                                 Choose a thumbnial...
                            </span></span> <span class="file-name">
                            
                        </span></label></div>
            </div>
        </div>
        <div class="field">
            <div class="control">
                <vue-html5-editor content="Start your story..." height="500"></vue-html5-editor>
            </div>
        </div>
        <div class="field is-grouped">
            <div class="control"><button id="submit-btn" class="button is-link">Post</button></div>
            <div class="control"><button type="button" class="button is-text">Cancel</button></div>
        </div>
    </form>
</div>
`;

exports[`PostForm.vue should render as edit mode. 1`] = `
<div class="section">
    <h1 class="title is-size-3">Edit Post</h1>
    <form class="form">
        <div class="field">
            <div class="control"><input type="text" name="title" placeholder="What's it about..." required="required" class="input"></div>
        </div>
        <div class="field">
            <div class="control">
                <div class="file has-name is-fullwidth"><label class="file-label"><input type="file" name="image" class="file-input"> <span class="file-cta"><span class="file-icon"><i class="fa fa-upload"></i></span> <span class="file-label">
                                 Choose a thumbnial...
                            </span></span> <span class="file-name">
                            
                        </span></label></div>
            </div>
        </div>
        <div class="field">
            <div class="control">
                <vue-html5-editor content="Start your story..." height="500"></vue-html5-editor>
            </div>
        </div>
        <div class="field is-grouped">
            <div class="control"><button id="submit-btn" class="button is-link">Post</button></div>
            <div class="control"><button type="button" class="button is-text">Cancel</button></div>
        </div>
    </form>
</div>
`;
